fix library navigation & focus issues

ronso0 opened the pull request:
These fixes are supposed to make the controller navigation (Up, Down, MoveFocus) behave exactly like using Up, Down and Tab on a keyboard.

See: [hints for testing this](https://github.com/mixxxdj/mixxx/pull/2599#issuecomment-608424968)

**fix `WLibrary->getActiveView()->hasFocus()`**
in DlgRecording, DlgAutoDJ, DlgAnalysis, DlgHidden and DlgMissing to prevent unnecessary calls to `setLibraryFocus()` before issuing any `[Library],Move...` command

**rework LibraryControl::emitKeyEvent()**
Any widget can receive `Tab`, but don't send key events if Mixxx isn't focused.

This also prevents lib focus getting stuck at Clear button:
In master, the focus can be moved to the Clear button with [Library],MoveFocus[Backward, Forward] (when there is an active search), but then, MoveFocusForward would do nothing because behind the scenes library control tests if the sidebar or tracks table have focus; if false it would try to re-focus the tracks table: focus the sidebar > emulate Tab > then send the MoveFocus Tab = we end up at the Clear button again. 
With MoveFocus, -1 it would previously jump between sidebar and Clear button.

ronso0 committed:
fix WLibrary->getActiveView()->hasFocus() for AutoDJ, Recording, Analysis, Hidden & Missing

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -482,16 +482,26 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
     }
-    if (!m_pLibraryWidget->hasFocus() && !m_pSidebarWidget->hasFocus()) {
+    if (!QApplication::focusWindow()) {
+        qDebug()<<"Mixxx window is not focused, don't send key events";
```
Is this worth a debug message after the fix is merged?

daschuer replied:
Do you want to return here?

ronso0 replied:
`return;` added

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -482,16 +482,26 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
+        qDebug()<<"Mixxx window is not focused, don't send key events";
+    }
+
+    Qt::Key key = static_cast<Qt::Key>(event.key());
```
is this cast necessary to for the compares below? 
If not, I would prefer to keep it as int.
I guess the QT api uses int, in case of values the do not have an enum for. 
The same considerations apply for our code.

ronso0 replied:
changed to `bool keyIsTab = event.key() == 16777217;`

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -482,16 +482,26 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
+    if (key != Qt::Key_Tab && !m_pSidebarWidget->hasFocus()
+            && !m_pLibraryWidget->getActiveView()->hasFocus()) {
         setLibraryFocus();
     }
```
do you want return her as well?

ronso0 replied:
nope. this is meant to catch Up / Down / L / R moves which are executed below.

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -482,16 +482,26 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
     // Send the event pointer to the currently focused widget
     for (auto i = 0; i < event.count(); ++i) {
-        QApplication::sendEvent(focusWidget, &event);
+        QApplication::sendEvent(QApplication::focusWidget(), &event);
```
This can be any widget .. can this become a infinite loop?

ronso0 replied:
hmm good point.
I'll restore `auto focusWidget = QApplication::focusWidget()` above.

Not an infinte loop, but multipe events could be sent to different widgets.
A widget might get disabled after the first event and consecutive events from the queue would be sent to another widget then. I don't see how that could happen from within librarycontrol but it's safer, of course.

daschuer commented on src/library/librarycontrol.cpp:
```diff
@@ -482,16 +482,26 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
+        setLibraryFocus();
     }
     // Send the event pointer to the currently focused widget
     for (auto i = 0; i < event.count(); ++i) {
```
What is this loop for? The send event will include the count. 
In which cases is count set?

ronso0 replied:
I was under the impression mulitple `MoveDown` events might be sent at once, but doesn't seem to be possible.
Even though I didn't introduce this I'll double check. Maybe we can remove it. Then my [commment above](https://github.com/mixxxdj/mixxx/pull/2599#discussion_r398891942) doesn't apply.

ronso0 replied:
I'll leave this as is.
`slotScrollVertical(double v)` and `slotMoveHorizontal(double v)` for example can send multiple key presses.

uklotzde commented:
I didn't notice any new issues, LGTM.

But I don't know how to test any of the fixed issues.

ronso0 commented:
> But I don't know how to test any of the fixed issues.

The fixes are supposed to make the controller navigation (Up, Down, MoveFocus) behave like if you'd use Up, Down and Tab on a keyboard.

For testing, focus the AutoDJ table and send an `[Library],MoveUp/Down` signals with a Trax knob or arrow buttons to move the selection. The log stream should only "Keypress: Down (or Up)" once and no artificial Tab events anymore, because the table now reports its focus state. Same applies to Recording and Analysis.

Another test:
type something in the searchbox, then try to navigate to the tracks table with `[Library],MoveFocus double` or `[Library],MoveFocusForward 1` or `[Library],MoveFocusBackward 1`. It should act as if you used the Tab key (whereas before the focus would be stuck at the Clear button).

ronso0 commented:
@uklotzde Thanks for the hint, I updated the description.

ronso0 commented:
Please see #2644 for testing all current lib navigation PRs combined.

Holzhaus commented on src/library/librarycontrol.cpp:
```diff
@@ -482,14 +482,25 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
+        return;
+    }
+
+    bool keyIsTab = event.key() == 16777217;
```
```suggestion
    bool keyIsTab = event.key() == static_cast<int>(Qt::Key_Tab);
```

ronso0 replied:
Alright, thx. Just learned that static_cast is converted at compile time.

Holzhaus commented on src/library/librarycontrol.cpp:
```diff
@@ -482,14 +482,25 @@ void LibraryControl::emitKeyEvent(QKeyEvent&& event) {
     }
-    if (!m_pLibraryWidget->hasFocus() && !m_pSidebarWidget->hasFocus()) {
+    if (!QApplication::focusWindow()) {
+        qDebug()<<"Mixxx window is not focused, don't send key events";
```
```suggestion
        qDebug() << "Mixxx window is not focused, don't send key events";
```

ronso0 replied:
Done.

ronso0 committed:
rework LibraryControl::emitKeyEvent()

Allow any widget to receive Tab key event, other keys are only sent
to sidebar or tracks table.

ronso0 commented:
Added the suggestions by @Holzhaus and squashed commits.
Ready

Holzhaus reviewed (commented):
LGTM, but someone else needs to do the manual test since my controller does not use the new library navigation COs.

ronso0 commented:
Somehow a **!** got lost from the code on the way.. fixed.

Be-ing commented:
I modified my controller script to test this and it works. Thanks for your patience.
